---
output:
  rmarkdown::html_document:
    highlight: pygments
    toc: false
    toc_depth: 3
    fig_width: 5
vignette: >
  %\VignetteIndexEntry{Basic manipulation of GenomicRanges}
  %\VignetteEngine{knitr::rmarkdown}
  %\VignetteEncoding[utf8]{inputenc}  
---

# Basic manipulation of *GRanges* objects

```{r GRanges-location-accessors}
seqnames(gr)
ranges(gr)
strand(gr)
```

The `granges()` function extracts genomic ranges without corresponding
metadata,

```{r granges-accessor}
granges(gr)
``` 

The `start()`, `end()`, `width()`, and `range` functions extract basic
interval characteristics,

```{r IRangesStuff}
start(gr)
end(gr)
width(gr)
```

The `mcols()` accessor extracts the metadata as a *DataFrame*,
 
```{r metadataAccess}
mcols(gr)
mcols(gr)$score
score(gr)
```

The lengths and other properties of the sequences containing the
ranges can (and should) be stored in the *GRanges* object. Formal
tracking of the sequence universe, typically the genome build, ensures
data integrity and prevents accidental mixing of ranges from
incompatible contexts. Assuming these data are of *Homo sapiens*, we
could add the sequence information like this:
 
```{r setSeqLengths}
seqinfo(gr) <- Seqinfo(genome="hg38")
```
The `Seqinfo()` function automatically loads the sequence information
for the specified `genome=` by querying the UCSC database.
 
And then retrieves as:
```{r setSeqLengths2}
seqinfo(gr)
```
 
Methods for accessing the `length` and `names` have
also been defined.
 
```{r names}
names(gr)
length(gr)
```

## Subsetting  *GRanges* objects

*GRanges* objects act like vectors of ranges, with the expected
vector-like subsetting operations available
 
```{r subset1}
gr[2:3]
```
 
A second argument to the `[` subset operator specifies which metadata
columns to extract from the *GRanges* object. For example,
 
```{r subset2}
gr[2:3, "GC"]
```

The `subset()` function provides an easy way to subset based on
attributes of the ranges and columns in the metadata. For example,
```{r subset3}
subset(gr, strand == "+" & score > 5, select = score)
```

Elements can also be assigned to the *GRanges* object.  This example
replaces the the second row of a *GRanges* object with the first row
of `gr`.
 
```{r assign1}
grMod <- gr
grMod[2] <- gr[1]
head(grMod, n=3)
```

There are methods to repeat, reverse, or select specific portions of
*GRanges* objects.
 
```{r other}
rep(gr[2], times = 3)
rev(gr)
head(gr,n=2)
tail(gr,n=2)
window(gr, start=2,end=4)
gr[IRanges(start=c(2,7), end=c(3,9))]
```

## Splitting and combining *GRanges* objects

THe `split()` function divides a *GRanges* into groups, returning a
*GRangesList*, a class that we will describe and demonstrate later.
 
```{r splitAppendGRanges}
sp <- split(gr, rep(1:2, each=5))
sp
```

We can split the ranges by metadata columns, like strand,

```{r splitByFormula}
split(gr, ~ strand)
```

The `c()` and `append()` functions combine two (or more in the case of
`c()`) *GRanges* objects.
 
```{r combine}
c(sp[[1]], sp[[2]])
```

The `stack()` function stacks the elements of a *GRangesList* into a
single *GRanges* and adds a column indicating the origin of each
element,
```{r stack}
stack(sp, index.var="group")
```

## Aggregating *GRanges* objects

Like other tabular data structures, we can aggregate *GRanges*
objects, for example,

```{r aggregate}
aggregate(gr, score ~ strand, mean)
```

The `aggregate()` function also supports a syntax similar to
`summarize()` from dplyr,

```{r aggregate2}
aggregate(gr, ~ strand, n_score = lengths(score), mean_score = mean(score))
```

Note that we need to call `lengths(score)` instead of `length(score)`
because `score` is actually a list-like object in the aggregation
expression.

## Exercises (5 min)

1. What are the chromosome lengths of the mouse `mm9` genome assembly?
2. Split the GRanges object `gr` by chromosome, and compute the mean GC content for each chromosome.
3. Install the TxDb package for the yeast (_Saccharomyces cerevisiae_) `sacCer3` genome assembly. How many genes are there on each chromosome?

