---
output:
  rmarkdown::html_document:
    highlight: pygments
    toc: false
    toc_depth: 3
    fig_width: 5
vignette: >
  %\VignetteIndexEntry{Downstream analysis}
  %\VignetteEngine{knitr::rmarkdown}
  %\VignetteEncoding[utf8]{inputenc}  
---

[DESeq2]: https://bioconductor.org/packages/DESeq2

# Downstream analysis

In this section we look into downstream analysis of gene expression data, and the way in which skills learned in working with _Bioconductor_ objects in one package translate to working with objects in other packages.

We start by loading the [DESeq2][] package, a very popular facility for analysing
differential expression of bulk RNA-seq data.

```{r, message=FALSE}
library("DESeq2")
```

The package requires count data like that in the `SummarizedExperiment` we have been working with, in addition to a `formula` describing the experimental design. We use the cell line as a covariate, and dexamethazone treatment as the main factor that we are interested in.

```{r}
dds <- DESeqDataSet(se, design = ~ cell + dex)
dds
```

The `dds` object can be manipulated very much like a `SummarizedExperiment`.

The DESeq workflow is summarized by a single function call, which performs statistical analysis on the data in the `dds` object.

```{r}
dds <- DESeq(dds)
```

A table summarizing measures of differential expression can be extracted from the object, and visualized or manipulated using commands we learned earlier today.

```{r}
results(dds)
```

